src/restLink.ts

```diff
@@ -530,9 +544,41 @@ export class RestLink extends ApolloLink {
         patchDeeper: RestLink.FunctionalTypePatcher,
       ) => {
         if (Array.isArray(data)) {
-          return data.map(d => patchDeeper(d, outerType, patchDeeper));
+          return data.map(
+            d =>
+              typeof d === 'object'
+                ? patchDeeper(d, outerType, patchDeeper)
+                : d,
+          );
         }
-        const subPatcher = table[outerType] || (result => result);
+
+        const subPatcher =
+          table[outerType] ||
+          (result => {
+            if (
+              !this.automaticallyInferTypenames ||
+              typeof result !== 'object'
+            ) {
+              return result;
+            }
+
+            if (!Array.isArray(result) && result.__typename === undefined) {
+              result.__typename = outerType;
+            }
+
+            const patched = Object.keys(result)
+              .filter(key => typeof data[key] === 'object')
+              .reduce((previousValue, key) => {
+                const typename = toUpperCamelCase(key);
```

Comment:
Does this mean you can use:
```graphql
query {
# […snipped…]
     foo @type()
  }
}
```

If yes, then we should document that in the comments or somewhere.

Additionally, this means that `uppercamelcase` should be a regular dependency not a `devDependency`!